src/utils/units.cpp

```diff
@@ -0,0 +1,115 @@
+//========================================================================================
+// (C) (or copyright) 2024. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#include "units.hpp"
+
+namespace ArtemisUtils {
+
+constexpr Real Msolar = 1.988416e33;
+constexpr Real AU = 1.495978707e13;
+constexpr Real Year = 31536000;
+constexpr Real parsec = 3.0857e18;
+constexpr Real Rjup = 6.991100e6;
+constexpr Real Mjup = 1.8982e30;
+
+Units::Units(ParameterInput *pin, std::shared_ptr<StateDescriptor> pkg) {
+  std::string unit_system_str = pin->GetOrAddString("artemis/units", "type", "scalefree");
+  if (unit_system_str == "scalefree") {
+    unit_system_ = UnitSystem::scalefree;
+  } else if (unit_system_str == "cgs") {
+    unit_system_ = UnitSystem::cgs;
+  } else {
+    PARTHENON_FAIL("Unit system not recognized! Valid choices are [scalefree, cgs]");
+  }
+
+  if (unit_system_ == UnitSystem::scalefree) {
+    length_ = 1.;
+    time_ = 1.;
+    mass_ = 1.;
+  } else {
+    std::string unit_specifier = pin->GetOrAddString("artemis", "specifier", "base");
+    if (unit_specifier == "base") {
+      length_ = pin->GetReal("artemis", "length");
+      time_ = pin->GetReal("artemis", "time");
+      mass_ = pin->GetReal("artemis", "mass");
+    } else if (unit_specifier == "ppd") {
+      length_ = AU;
+      mass_ = Msolar;
+      time_ = Year / (2. * M_PI);
+    } else {
+      PARTHENON_FAIL("Unit specifier not recognized!");
+    }
+  }
+
+  // Remaining conversion factors
+  energy_ = std::pow(length_, 2) * mass_ * std::pow(time_, -2);
+  number_density_ = std::pow(length_, -3);
```

Comment:
Conversion factors for all the other compound quantities like `pressure`, `cv`, opacity, conductivity, viscosity?

Reply:
I see below that you just compute those conversion on the fly. Not sure why some are computed on the fly and some are stored, but not a big deal

Reply:
I tried to store the factors the need `std::pow` but calculate the others. It's just a choice I made.